File: data/translated_projects/commons-fileupload/src/main/org/apache/commons/fileupload/util/mime/QuotedPrintableDecoder.py
from __future__ import annotations
from io import StringIO
import io
from io import BytesIO
import typing
from typing import *
import java # type: ignore
from src.main.org.apache.commons.fileupload.java_handler import *
class QuotedPrintableDecoder(metaclass=StaticFieldRedirector):
    @staticmethod
    def decode(data: typing.List[int], out: typing.Union[io.BytesIO, io.StringIO, io.BufferedWriter]) -> int:
        off = 0
        length = len(data)
        end_offset = off + length
        bytes_written = 0

        while off < end_offset:
            ch = data[off]
            off += 1

            if ch == ord('_'):
                out.write(b' ')
            elif ch == ord('='):
                if off + 1 >= end_offset:
                    raise IOError("Invalid quoted printable encoding; truncated escape sequence")

                b1 = data[off]
                off += 1
                b2 = data[off]
                off += 1

                if b1 == ord('\r'):
                    if b2 != ord('\n'):
                        raise IOError("Invalid quoted printable encoding; CR must be followed by LF")
                else:
                    c1 = QuotedPrintableDecoder.__hexToBinary(b1)
                    c2 = QuotedPrintableDecoder.__hexToBinary(b2)
                    out.write(bytes([(c1 << QuotedPrintableDecoder.__UPPER_NIBBLE_SHIFT) | c2]))
                    bytes_written += 1
            else:
                out.write(bytes([ch]))
                bytes_written += 1

        return bytes_written
    @staticmethod

    def __hexToBinary(b: int) -> int:

        idMapPyToJ = JavaHandler.valueToObject(dict(), "Map")
        translatedArg0 = JavaHandler.valueToObject(b, "byte", idMapPyToJ)
        idMapJToPy = dict()
        try:
            val = JavaHandler.mapping(QuotedPrintableDecoder.javaClz.hexToBinary(translatedArg0), id_map=idMapJToPy)
            return val
        except:
            raise JavaHandler.mapping(java.type("org.apache.commons.fileupload.ExceptionHandler").ERR)
        finally:
            pass

            JavaHandler.mapping(translatedArg0, idMapJToPy, b)
    def __init__(self) -> None:

        idMapPyToJ = JavaHandler.valueToObject(dict(), "Map")
        
        idMapJToPy = dict()
        try:
            self.javaObj = QuotedPrintableDecoder.javaClz()
            self.javaObj.setPythonObject(self)
            self.javaObj.jToPy(idMapJToPy)
        except:
            raise JavaHandler.mapping(java.type("org.apache.commons.fileupload.ExceptionHandler").ERR)
        finally:
            pass
            
    javaClz = java.type("org.apache.commons.fileupload.util.mime.QuotedPrintableDecoder")
    def __getattr__(self, name):
        try:
            return object.__getattribute__(self, name)
        except AttributeError:
            pass
        return JavaHandler.mapping(getattr(self.javaClz, StaticFieldRedirector.unmangle_name(name)))

File: logs/glue/commons-fileupload/src/main/java/org/apache/commons/fileupload/util/mime/QuotedPrintableDecoder.java
package org.apache.commons.fileupload.util.mime;

import java.io.IOException;
import java.io.OutputStream;
import org.apache.commons.fileupload.ContextInitializer;
import org.apache.commons.fileupload.ExceptionHandler;
import org.apache.commons.fileupload.IntegrationUtils;
import org.graalvm.polyglot.PolyglotException;
import org.graalvm.polyglot.Value;

public final class QuotedPrintableDecoder {
  private transient Value obj =
      IntegrationUtils.createDefaultPythonObject(
          ContextInitializer.getPythonClass(
              "main/org/apache/commons/fileupload/util/mime/QuotedPrintableDecoder.py",
              "QuotedPrintableDecoder"));
  public static final int UPPER_NIBBLE_SHIFT = Byte.SIZE / 2;

  public Value getPythonObject() {
    return obj;
  }

  public void setPythonObject(Value obj) {
    this.obj = obj;
  }

  public static int decode(byte[] data, OutputStream out) throws IOException {
    //
    // int off = 0;
    // int length = data.length;
    // int endOffset = off + length;
    // int bytesWritten = 0;
    //
    // while (off < endOffset) {
    // byte ch = data[off++];
    //
    // if (ch == '_') {
    // out.write(' ');
    // } else if (ch == '=') {
    // if (off + 1 >= endOffset) {
    // throw new IOException(
    // "Invalid quoted printable encoding; truncated escape sequence");
    // }
    //
    // byte b1 = data[off++];
    // byte b2 = data[off++];
    //
    // if (b1 == '\r') {
    // if (b2 != '\n') {
    // throw new IOException(
    // "Invalid quoted printable encoding; CR must be followed by LF");
    // }
    // } else {
    // int c1 = hexToBinary(b1);
    // int c2 = hexToBinary(b2);
    // out.write((c1 << UPPER_NIBBLE_SHIFT) | c2);
    // bytesWritten++;
    // }
    // } else {
    // out.write(ch);
    // bytesWritten++;
    // }
    // }
    //
    // return bytesWritten;
    //

    Value idMapJToPy = IntegrationUtils.mapToPython(new java.util.HashMap());
    Value translatedArg0 = IntegrationUtils.mapToPython(data, idMapJToPy);
    Value translatedArg1 = IntegrationUtils.mapToPython(out, idMapJToPy);
    java.util.Map idMapPyToJ = new java.util.HashMap();
    try {
      int val =
          (int)
              IntegrationUtils.valueToObject(
                  ContextInitializer.getPythonClass(
                          "main/org/apache/commons/fileupload/util/mime/QuotedPrintableDecoder.py",
                          "QuotedPrintableDecoder")
                      .invokeMember("decode", translatedArg0, translatedArg1),
                  "int",
                  idMapPyToJ);
      return val;
    } catch (PolyglotException e) {
      throw (IOException) ExceptionHandler.handle(e, "QuotedPrintableDecoder.decode");
    } finally {
      IntegrationUtils.valueToObject(translatedArg0, "byte[]", idMapPyToJ, byte.class, data);
      IntegrationUtils.valueToObject(translatedArg1, "OutputStream", idMapPyToJ, out);
    }
  }

  public static int hexToBinary(final byte b) throws IOException {

    try {

      final int i = Character.digit((char) b, 16);
      if (i == -1) {
        throw new IOException("Invalid quoted printable encoding: not a valid hex digit: " + b);
      }
      return i;

    } catch (Throwable ExceptionObjectForCaching) {
      ExceptionHandler.ERR = ExceptionObjectForCaching;
      throw ExceptionObjectForCaching;
    }
  }

  public QuotedPrintableDecoder() {}

  public java.util.Map pyToJ() {
    java.util.Map idMap = new java.util.HashMap();
    return pyToJ(idMap);
  }

  public java.util.Map pyToJ(java.util.Map idMap) {

    return idMap;
  }

  public Value jToPy() {
    Value idMap = IntegrationUtils.mapToPython(new java.util.HashMap());
    return jToPy(idMap);
  }

  public Value jToPy(Value idMap) {

    this.obj.invokeMember("__setattr__", "javaObj", this);

    return idMap;
  }
}
